### PR TITLE
refactor(forEach): change to for-of with iterable

### LINKED ISSUE #909: Extend *foreach syntax to support ES6 data structures
~~This might be a pedantic request, but shouldn't `*foreach` microsyntax use `of` instead of `in`, since a for-in loop (`for (item in items)`) iterates an array's keys rather than its values?~~

This issue has expanded to include more than array iteration. [One of my comments on this thread](https://github.com/angular/angular/issues/909#issuecomment-78165514) describes one possibility for making the `*foreach` syntax more flexible.

### PR DESCRIPTION
**Pull request by PatrickJS**
Tests are passing. I also used `git mv` to preserve the git history. I changed references with the name array to iterable
closes #909

### CODE REVIEW

**Comment by vicb** [comment]
@gdi2290 the Travis error comes from the benchmarks. infinite_scroll uses `ForEach/For` and needs to be updated

**Comment by gdi2290** [comment]
@vicb thanks I couldn't make out what the Travis error was. Give me a second

**Comment by gdi2290** [comment]
@vicb Dart doesn't have a for-of so how do we change the interface for Dart users to use for-in. Also where is the bug for the es6 version (for...in)?

**Comment by gdi2290** [comment]
:/ I'm doing something wrong with my Dart version
```dart
void iterateListLike(iter, fn(item)) {
  if (iter is List) {
    for (var item = 0; item < iter.length; ++item) {
      fn(iter[item]);
    }
  } else {
    for (var item in iter) {
      fn(item);
    }
  }
}
```

**Comment by gdi2290** [comment]
@mhevery I rebased and fixed the merge conflicts

**Comment by gdi2290** [comment]
is there a better way to rebase on github? I always force push up changes after rebase but that also blows up any commits that were made in the files previously :/

File: modules/angular2/test/directives/for_spec.js [hunk 1]
```diff
@@ -124,8 +124,8 @@ export function main() {
       });
     }));
 
-    it('should iterate over an array of objects', () => {
-      compileWithTemplate('<ul><li template="foreach #item in items">{{item["name"]}};</li></ul>').then((pv) => {
+    it('should iterate over an array of objects', inject([AsyncTestCompleter], (async) => {
```
Comment (gdi2290) [inline modules/angular2/test/directives/for_spec.js]: I also fixed the build in this commit
[foreach_spec.js#L127])https://github.com/angular/angular/blob/afda43dc0281e3671d6df3218749d5050d47f188/modules/angular2/test/directives/foreach_spec.js#L127)

**Comment by mhevery** [comment]
In general I like this commit. You did make the change detection slower however. Those parts need to be reverted, but the rest can go. Please ping this thread once ready.

**Comment by gdi2290** [comment]
The commit was squash without other changes. I also preserved the git history during renaming. There needs to be Iterator facade next and then a perf PR to avoid `polymorphic` code for when it's an iterable

**Comment by caitp** [comment]
there are only a handful of iterables most people are going to use, it's not likely to result in IC misses if the path is only taken when the object is known statically to be iterable

**Comment by gdi2290** [comment]
gg I have to update the TodoMVC example to use `For`

**Comment by gdi2290** [comment]
@mhevery PR is all green now with correct rebase and renames tracked by git

**Comment by gdi2290** [comment]
@mhevery I rebase'd Master again so everything is green. If you have any suggestions or changes feel free to ping me again about it

**Comment by mhevery** [comment]
b61b8d60b7cfd10786ccd9ba4071738d4f2bf15a

**Pull request closed without merging**